RecoEcal/EgammaClusterAlgos/src/SCEnergyCorrectorSemiParm.cc

```diff
@@ -1,324 +1,383 @@
-
-    if (nullptr != recHitsEB) {
-      for (EcalRecHitCollection::const_iterator it = recHitsEB->begin(); it != recHitsEB->end(); ++it) {
-        if (it->energy() > eThreshold_)
-          nHitsAboveThreshold_++;
-      }
-    }
+void SCEnergyCorrectorSemiParm::setEventSetup(const edm::EventSetup& es) {
+  caloTopo_ = es.getHandle(caloTopoToken_);
+  caloGeom_ = es.getHandle(caloGeomToken_);
 
-    if (nullptr != recHitsEE) {
-      for (EcalRecHitCollection::const_iterator it = recHitsEE->begin(); it != recHitsEE->end(); ++it) {
-        if (it->energy() > eThreshold_)
-          nHitsAboveThreshold_++;
-      }
-    }
+  regParamBarrel_.setForests(es);
+  regParamEndcap_.setForests(es);
+
+  if (isPhaseII_) {
+    hgcalShowerShapes_.initPerSetup(es);
   }
 }
 
-//--------------------------------------------------------------------------------------------------
-std::pair<double, double> SCEnergyCorrectorSemiParm::getCorrections(const reco::SuperCluster &sc) const {
-  std::pair<double, double> p;
-  p.first = -1;
-  p.second = -1;
+void SCEnergyCorrectorSemiParm::setEvent(const edm::Event& event) {
+  event.getByToken(tokenEBRecHits_, recHitsEB_);
+  if (!isPhaseII_) {
+    event.getByToken(tokenEERecHits_, recHitsEE_);
+  } else {
+    event.getByToken(tokenHgcalRecHits_, recHitsHgcal_);
+    hgcalShowerShapes_.initPerEvent(*recHitsHgcal_);
+  }
+  if (isHLT_ || isPhaseII_) {
+    //note countRecHits checks the validity of the handle and returns 0
+    //if invalid so its okay to call on all rec-hit collections here
+    nHitsAboveThresholdEB_ = countRecHits(recHitsEB_, hitsEnergyThreshold_);
+    nHitsAboveThresholdEE_ = countRecHits(recHitsEE_, hitsEnergyThreshold_);
+    nHitsAboveThresholdHG_ = countRecHits(recHitsHgcal_, hitsEnergyThreshold_);
+  }
+  if (!isHLT_) {
+    event.getByToken(tokenVertices_, vertices_);
+  }
+}
 
-  // protect against HGCal, don't mod the object
-  if (EcalTools::isHGCalDet(sc.seed()->seed().det()))
-    return p;
+std::pair<double, double> SCEnergyCorrectorSemiParm::getCorrections(const reco::SuperCluster& sc) const {
+  std::pair<double, double> corrEnergyAndRes = {-1, -1};
 
-  const reco::CaloCluster &seedCluster = *(sc.seed());
-  const bool iseb = seedCluster.hitsAndFractions()[0].first.subdetId() == EcalBarrel;
-  const EcalRecHitCollection *recHits = iseb ? rechitsEB_.product() : rechitsEE_.product();
+  const auto regData = getRegData(sc);
+  if (regData.empty()) {
+    //supercluster has no valid regression, return default values
+    return corrEnergyAndRes;
+  }
+  const auto& regParam = getRegParam(sc.seed()->seed());
 
-  const CaloTopology *topo = calotopo_.product();
+  double mean = regParam.mean(regData);
+  double sigma = regParam.sigma(regData);
 
-  const double raw_energy = sc.rawEnergy();
-  const int numberOfClusters = sc.clusters().size();
+  double energyCorr = mean * sc.rawEnergy();
```

Comment:
so Slava here is where I fixed the rounding error.  In the original sc.rawEnergy() which is a double is stored in  float which is then used in this calculation. By putting   ` double energyCorr = mean * static_cast<float>(sc.rawEnergy());` I got the energies closer and the conversion had no differences after that. 

Note I dont propose to update the PR with this cast as we should be able to tolerate these rounding differences.